crc/riscv64/crc_riscv64_dispatcher.c

```diff
@@ -0,0 +1,252 @@
+
+extern uint64_t
+crc64_ecma_norm_vclmul(uint64_t, const unsigned char *, uint64_t);
+
+extern uint64_t
+crc64_iso_refl_vclmul(uint64_t, const unsigned char *, uint64_t);
+
+extern uint64_t
+crc64_iso_norm_vclmul(uint64_t, const unsigned char *, uint64_t);
+
+extern uint64_t
+crc64_jones_refl_vclmul(uint64_t, const unsigned char *, uint64_t);
+
+extern uint64_t
+crc64_jones_norm_vclmul(uint64_t, const unsigned char *, uint64_t);
+
+extern uint64_t
+crc64_rocksoft_refl_vclmul(uint64_t, const unsigned char *, uint64_t);
+
+extern uint64_t
+crc64_rocksoft_norm_vclmul(uint64_t, const unsigned char *, uint64_t);
+
+DEFINE_INTERFACE_DISPATCHER(crc16_t10dif)
+{
+#if HAVE_RVV && HAVE_ZBC && HAVE_ZVBC && HAVE_ZVBB
+        struct riscv_hwprobe _probe = INIT_PROBE_STRUCT();
+        syscall(__NR_riscv_hwprobe, &_probe, 1, 0, NULL, 0);
+        unsigned long auxval = getauxval(AT_HWCAP);
+        if (auxval & HWCAP_RV('V') && _probe.value & HWCAP_RV_EXT("ZVBC") &&
+            _probe.value & HWCAP_RV_EXT("ZVBB") && _probe.value & HWCAP_RV_EXT("ZBC")) {
+                return crc16_t10dif_vclmul;
+        }
+#endif
+        return crc16_t10dif_base;
+}
+
+DEFINE_INTERFACE_DISPATCHER(crc16_t10dif_copy)
+{
+#if HAVE_RVV && HAVE_ZBC && HAVE_ZVBC && HAVE_ZVBB
+        struct riscv_hwprobe _probe = INIT_PROBE_STRUCT();
+        syscall(__NR_riscv_hwprobe, &_probe, 1, 0, NULL, 0);
+        unsigned long auxval = getauxval(AT_HWCAP);
+        if (auxval & HWCAP_RV('V') && _probe.value & HWCAP_RV_EXT("ZVBC") &&
+            _probe.value & HWCAP_RV_EXT("ZVBB") && _probe.value & HWCAP_RV_EXT("ZBC")) {
+                return crc16_t10dif_copy_vclmul;
+        }
+#endif
+        return crc16_t10dif_copy_base;
+}
+
+DEFINE_INTERFACE_DISPATCHER(crc32_ieee)
+{
+#if HAVE_RVV && HAVE_ZBC && HAVE_ZVBC && HAVE_ZVBB
+        struct riscv_hwprobe _probe = INIT_PROBE_STRUCT();
+        syscall(__NR_riscv_hwprobe, &_probe, 1, 0, NULL, 0);
+        unsigned long auxval = getauxval(AT_HWCAP);
+        if (auxval & HWCAP_RV('V') && _probe.value & HWCAP_RV_EXT("ZVBC") &&
+            _probe.value & HWCAP_RV_EXT("ZVBB") && _probe.value & HWCAP_RV_EXT("ZBC")) {
+                return crc32_ieee_norm_vclmul;
+        }
+#endif
+        return crc32_ieee_base;
+}
+
+DEFINE_INTERFACE_DISPATCHER(crc32_iscsi)
+{
+#if HAVE_RVV && HAVE_ZBC && HAVE_ZVBC && HAVE_ZVBB
+        struct riscv_hwprobe _probe = INIT_PROBE_STRUCT();
+        syscall(__NR_riscv_hwprobe, &_probe, 1, 0, NULL, 0);
+        unsigned long auxval = getauxval(AT_HWCAP);
+        if (auxval & HWCAP_RV('V') && _probe.value & HWCAP_RV_EXT("ZVBC") &&
+            _probe.value & HWCAP_RV_EXT("ZVBB") && _probe.value & HWCAP_RV_EXT("ZBC")) {
+                return crc32_iscsi_refl_vclmul;
+        }
+#endif
+        return crc32_iscsi_base;
+}
+
+DEFINE_INTERFACE_DISPATCHER(crc32_gzip_refl)
+{
+#if HAVE_RVV && HAVE_ZBC && HAVE_ZVBC && HAVE_ZVBB
+        struct riscv_hwprobe _probe = INIT_PROBE_STRUCT();
```

Comment:
All these lines (from 131 to 135) are always the same throughout this file. Would be better to move it to a separate function.

Reply:
Looks tidier, yes